ohnosequences/nice-sbt-settings
# Simplify release process

After #35 some things in the release will be broken. But with the new version plugin things can be simplified a lot: 
- no `version.sbt`, no writing to files and committing it during release
- to start release process you choose: `major`, `minor`, `bugfix`, `milestone`, or `releaseCandidate` and the version is set automatically (and no need in the "next" version, as now we have only current version)

---

Here is a draft of the simplified release process:
1. pre-release checks (before any changes are introduced). if any of these checks don't pass, there's no point in even starting the release process.
   - [x] `checkGit`:
     - no uncommited changes
     - check that there's no tag with the release version name already
     - update remote and check that current branch is up to date (being ahead is ok)
     - check GitHub credentials
   - [x] `checkCodeNotes`: TODO/FIXME list (ask to continue if there are some)
   - [x] `checkDependencies`: 
     - no snapshot dependencies
     - check dependency updates (ask to continue if there are some)
     - run `update`
   - [x] run _non-release_ tests (`test in Test`)
   - [x] `prepareReleaseNotesAndTag`:
     - check release notes for corresponding version (show and confirm)
     - rename them and commit if needed
     - create git tag for the release version
2. **reload**. after this `version` setting has the release version used by all further steps.
3. package and publish artifacts
4. run release-only tests (they may require the artifact)
5. publish GitHub release:
   - push git tag
   - run `releaseOnGithub`
6. handle external documentation (should be optional):
   - clean/generate literator source docs, commit and push it
   - generate API docs and push them to GitHub-pages

---

If steps 3 or 4 fail, we can always checkout the newly made tag and delete it (so the release process is aborted without any changes introduced). After step 5 the tag became public and any changes should go in the next release.
